### ISSUE TITLE
High Frequency requests to controller methods cause unnecessary and excessive UPDATE calls to the DB to extend the session state 

### ISSUE THREAD

**Issue by precisioninfinity** [aspnet/AspNetSessionState#40]
Even though controller methods don't use session state the library still seems to wire up to an end request at which point it fires a call to ResetItemTimeout.  For something like controller methods which can get hit hundreds of times in a short time period and are supposed to be mostly stateless, it's still generating a SQL UPDATE for each one of those hits, which causes massive locking waits backups and is exhausting the worker threads on Azure SQL which is an extremely limited resource.  In our case it was causing users to get sql timeout errors and an error that said 'The request limit for the elastic pool is [x] and has been reached.'

Even with aspnet:AllowConcurrentRequestsPerSession set to true it's still problematic.

We resolved the issue by extending the SqlSessionStateProviderAsync class with the following, which someone could easily turn into a pull request if they like.  Posting here in case it helps someone else.

`    public class ControllerSafeSqlSessionStateProviderAsync : SqlSessionStateProviderAsync
    {
        private readonly string cacheKeyPrefix =
            $"SessionLastSet-{nameof(ControllerSafeSqlSessionStateProviderAsync)}.{nameof(ResetItemTimeoutAsync)}";

        private readonly int IgnoreSuccessiveCallsToResetTimeoutInSeconds =
            WebConfigurationManager.AppSettings["aspnet:SecondsToIgnoreSuccessiveSessionTimeoutReset"].ParseValueOrNull<int>() ?? 30;

        public override async Task ResetItemTimeoutAsync(HttpContextBase context, string id, CancellationToken cancellationToken)
        {
            // this gets called way more than it needs to...
            // each request (which for webapi controller calls is supposed to be quick)
            // is extending the session at the end of the request
            // (even on controllers which don't always use the session)
            // these excessive calls sare 
            // adding significant locking waits to the db server
            // because 500 webapi requests in a minute 
            // don't all need to extend the session timeout, we ignore
            // any successive ones in x seconds
            var cacheKey = $"{cacheKeyPrefix}.{id}";
            var result = MemoryCache.Default.Get(cacheKey) as DateTime?;

            // if the value was set in the last x seconds then... seriously?... you are calling us too much
            if (result.HasValue && DateTime.Now > result?.AddSeconds(-1 * IgnoreSuccessiveCallsToResetTimeoutInSeconds))
                return; // return quickly to not hold up the request

            if (!cancellationToken.IsCancellationRequested)
            {
                await base.ResetItemTimeoutAsync(context, id, cancellationToken); // purposely passing null because the base class impl doesn't use it and it might be gone
                MemoryCache.Default.Set(cacheKey, DateTime.Now, DateTime.Now.AddSeconds(IgnoreSuccessiveCallsToResetTimeoutInSeconds));
            }
        }
    }`

**Comment by wnayes** [comment]
I think the problem is in `AcquireStateAsync` when the module calls `ResetItemTimeoutAsync` for any request that does not use session state.

https://github.com/aspnet/AspNetSessionState/blob/master/src/SessionStateModule/SessionStateModuleAsync.cs#L416

```
// Still need to update the sliding timeout to keep session alive.
// There is a plan to skip this for perf reason.  But it was postponed to
// after Whidbey.
await _store.ResetItemTimeoutAsync(_rqContext, _rqId, GetCancellationToken());
```

It would be nice if a setting could be added to disable this behavior.

**Comment by precisioninfinity** [comment]
Yeah that's what I was thinking too.  Instead of disabling it completely just skipping it if it's been called for that _rqId within the past 30 seconds would be fine.  And maybe a config value for the 60 seconds to override it.  I think the sliding session timeout is 20 minutes by default, so you don't need to keep updating the sliding timeout if when this method is called 100 times within a 30 second window.  Which is basically what the code above does.  Already tested on production and made a huge difference.  It's a problem we noticed a year ago and finally isolated to this only recently.

**Comment by precisioninfinity** [comment]
I forgot to mention the problem this caused and why we had to add the workaround. If there's a high number of requests that come in, say rapidly to controller methods, then there can be so many DB worker threads from this that the DB wait timeouts get really high while the system is waiting for an available DB worker thread. Standard Azure SQL really limits the worker threads, I think to like only 50.  One user can cause hundreds of simultaneous controller method hits in an MVC app easily. Even though the controller methods might be done right and return fast, the overhead of this module being DB dependent really slows it down.
